src/pages/posts/comparing-rust-smart-contract-storage-approaches.md

```diff
@@ -0,0 +1,515 @@
+
+In the early days of Cosmos, in order to make smart contracts, you needed to hard-code the contracts into your validator runtime. This meant that simple smart contract changes essentially required a hard-fork of the network. This is fine for app-specific chains with well-staffed teams that understand the whole stack and operate the whole network, but made Cosmos an impractical choice for simpler apps that didn't need a whole chain of their own.
+
+CosmWasm aims to split the difference between this app-specific chain model and the one-chain, many-contracts model of blockchains like Ethereum (or NEAR or Soroban). It's a plugin for CosmosSDK that makes it easy to add support for generic smart contracts on whatever blockchain you want to build. Like NEAR and Soroban, CosmWasm compiles smart contracts to WebAssembly, aka Wasm. This gives these smart contracts the same characteristics you're used to from other blockchains: they can be upgraded without hard-forking the network; they each run safely in their own sandbox, unable to access protected parts of the validator runtime host environment.
+
+So CosmWasm does a lot more than something like near-sdk-rs. near-sdk-rs gets to focus just on helping you write smart contracts for NEAR, but CosmWasm also needs to define the whole "backend" part that plugs the Wasm-host logic into the larger CosmosSDK. It actually handles [even more than that](https://github.com/Web3-Builders-Alliance/ClusterOneCodeJournal.W22MTW.Chad/blob/main/standards.md).
+
+And, aside from all this, CosmWasm also encourages many more conventions around smart contract authoring. If you start with the [recommended template](https://github.com/CosmWasm/cw-template), you'll end up with more than five files instead of just `src/lib.rs`, splitting your contract logic into _state_, _messages_, _errors_, and then the actual contract logic.
+
+For the sake of side-by-side comparison, I've [consolidated all of this into one file](https://github.com/AhaLabs/cosmwasm-single-file).
+
+Without further ado, here it is, the "simple" CosmWasm incrementer:
+
+```rust
+use cosmwasm_schema::{cw_serde, QueryResponses};
+use cosmwasm_std::{
+    entry_point, to_binary, Binary, Deps, DepsMut, Env, MessageInfo, Response, StdError, StdResult,
+};
+use cw_storage_plus::Item;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct State {
+    pub count: i8,
+}
+
+pub const STATE: Item<State> = Item::new("state");
+
+#[cw_serde]
+pub struct InstantiateMsg {
+    pub count: i8,
```

Comment:
Do you need to include data here at all? Then `derive(Default)` `State` and then 
```rust
    let state = State { count: msg.count };
```
becomes

```
    let state = State::default();
```

Reply:
Yeah, in this case, you could definitely do that. I'm not sure if I should update the post, though, since what I did here is the closest to what most CosmWasm contracts seem to use.

Reply:
I think you meant that `let state = State::default();` would happen _in_ `initialize`, yes? In which case CosmWasm is still a little bit different than NEAR/Soroban, since NEAR/Soroban allow having "no data stored until it's mutated" behavior. 

I don't think there's a way to avoid setting `STATE` within `INITIALIZE`, or at least none of the examples I've seen have pushed this as an option.

Avoiding the `msg.count` parameter to `initialize` seems like a smaller point.